### ISSUE TITLE
forbid-prop-types / Allow PropTypes.any in one case

### ISSUE THREAD

**Issue by hugomallet** [yannickcr/eslint-plugin-react#2520]
Hi,

I would like to avoid defining proptypes of props that are not directly used by a component but are just passed through. Example :
```js
// Parent.js
// In this case p2 is just passed through
const Parent = ({ p1, p2 }) => (
    <div><Child p2={p2} /> {p1}</div>
);
```

> I have already posted an issue in prop-types project that you shoud read to understand :
> https://github.com/facebook/prop-types/issues/305
> The result of the previous issue is that the feature i need should be at lint time not run time. This is why i'm here ;-)

As Parent does not need to know the type of p2, i would like to be able to define it as PropTypes.any().

But to avoid to have to disable the rule each time and to avoid later issues (for example if the prop is finally used in a later update) it would be great if the rule would have this behaviour : 
PropTypes.any() is allowed while the prop is only passed to a child component.

I don't know if it is feasible.

**Comment by ljharb** [comment]
I don’t think it is - my opinion there is the same as here, that the child component should be properly exposing its propTypes, or that a code comment suffices (including an eslint override).

**Comment by hugomallet** [comment]
@ljharb What would you recommend as a general rule ? disabling eslint rule on the line ? import child's proptypes (what if you can't) ?

**Comment by ljharb** [comment]
If you can, do; if you can’t, then personally I’d make my own version of what their propTypes should be, and import/use that; if you do in fact need an “any” then an eslint override comment.

**Comment by hugomallet** [comment]
Thanks. My goal is to avoid passing too much time defining (and re-re-re-defining) types. Our discussion points me out that what i need is maybe to avoid types more generally (proptypes and linting types). I am not a "type"script fan. Will think about that.

I keep this open for now. Let me know if any other idea that could simplify this comes to head :)

**Comment by ljharb** [comment]
I'm not a TS fan either, but propTypes are very critical (and more powerful in most ways than a type system). I'd suggest trying very hard to ensure everything has rigorous propTypes defined.

**Comment by hugomallet** [comment]
Could you develop or point me out an article that explains why they are very critical and powerful ?

**Comment by ljharb** [comment]
No, I'm not aware of any. Power, because they can, for example, target "integer", which TS can not - see https://npmjs.com/airbnb-prop-types for many examples.

**Comment by hugomallet** [comment]
Why component props should have types but function arguments not ? Is it only for "documentation" ?

**Comment by ljharb** [comment]
That's not a thing that's possible in JS from outside a function, so it's a tangential issue.

**Comment by hugomallet** [comment]
I was just trying to understand why component props should have their types verified, while in JS in general types are not verified.

**Comment by ljharb** [comment]
All function arguments should be verified in JS; just with runtime type checks. PropTypes are a DSL for those type checks, but in a way that easily strips them out in production.

**Comment by hugomallet** [comment]
I mean what the difference between these :
```jsx
// 1. Component
const ComponentLogger = (props) => {
    console.log(props.message);
    return null;
};

<ComponentLogger message="hello" />

// 2. Function
const functionLogger = (message) => {
    console.log(message);
    return null;
};

functionLogger('hello');
```

Are you saying that both should use PropTypes ?

**Comment by ljharb** [comment]
The first absolutely should. The second, since it's not a React component, should be something like this:
```js
const functionLogger = (message) => {
    if (typeof message !== 'string') {
      throw new TypeError('message must be a string');
    }
    console.log(message);
    return null;
};
```
